### ISSUE TITLE
Multiple Questions

### ISSUE THREAD

**Issue by JPhilipp** [Unity-Technologies/ml-agents#1572]
Thanks for the great ML-Agents! I have so many questions that I figured posting all at once is better this time.

- Am I right in assuming that existential negative rewards per step (usually used to reward faster accomplishments) aren't really needed when the agent or academy has a resetting max steps anyway, as it means agents achieving positive rewards faster are rewarded anyway in sum when the round ends?

- How does one approach situations where the reward is much delayed from a singular original action, say a cannon learning to orient itself towards shooting a target, but the bullet travels rather slowly and would only hit and miss after 1 second? For another example, if the agent's task is to stack up 10 differently-shaped rigidbodies as high as possible by using random positions and rotations, but where the height is only observed after 5 seconds to accomodate for stable towers.

- Is it appropriate to train the same brain in different environments of increasing complexity (without specifically using curriculums), e.g. first learn the sphere to roll towards targets, then later introducing walls so it can learn to jump etc.?

- The brains of the visual, camera-based examples provided with ML-Agents seem to be setup in a way that they don't actually yet have the camera added, but instead have a vector observation of 1, is that meant to then be changed by us to add the camera and set the observation to 0?

- Is there an example of a ready-trained, successful camera (not rayPerception) observation-based agent, and if not, could you include one?

- It would be super nice if in the textual examples background explanations, there was a chapter for each on design considerations. E.g. why the specific rayPerception angles were chosen, why it uses or doesn't use curiosity, how long it took to train them succesfully, what other parameters and setup considerations were used.

- Is anything on the roadmap for the future for training the setup values themselves via ML, e.g. an ML figuring out on its own the optimum number of rayPerception angles, or optimizing its batch_size and buffer_size settings etc.? (I find that sometimes setting these values is a big part of the human challenge, though I guess meta-ML'ing these could be tough for current computing powers as it takes so many more tries.)

- When is setting up multiple agents-with-same-environment in Unity beneficial, when isn't it -- e.g. the 12 platforms in the 3DBall balancer example, instead of using just 1 -- and how does one best figure out the amount of mirroring environments to use?

- It might be nice if there was a class like rayPerception but it takes its transform direction as basis, and then spreads its lasers out in more of a forward-circular fashion.

- How do I pick the ideal rayPerception max ray distance? Is it better for it to see further, or should I use rather close distances? (I'm not having too much luck with my agent training successes, in case it's not obvious yet :))

- Would it make sense to have a rayPerception type class but where it also collects the surface normals it hits, so that it's easier for the agent to differentiate between e.g. ground, ramp, and walls (without each of those getting specific tags)?

- What does the "IL" in the example scene names mean, does that stand for Imitiation Learning?

- In what types of situations would you suggest to use_curiosity, and when not? And how does one come up with a reasonable value for curiosity_strength and curiosity_enc_size?

- It would be cool if you could add a car, WheelCollider-based example, where each of the 4 wheels' motorTorque is somehow individually managed to some sort of success (I'm trying with ramps myself, rewarding for wheels not isGrounded as well as forward velocity, trying to train towards spectacular stunts).

- What's the current roadmap for ML-Agents?

Thanks so much!

**Comment by MarcoMeter** [comment]
I answered a few questions...


> Am I right in assuming that existential negative rewards per step [...] aren't really needed...

Defining reward signals is very dependent on the underlying environment. So there is no general right or wrong.

> How does one approach situations where the reward is much delayed from a singular original action...

Delayed rewards are a major challenge of reinforcement learning. In the case of running into this issue, I'd try to add more details to the observation space. So that for example the cannon can see the bullet flying.

> Is it appropriate to train the same brain in different environments of increasing complexity...

It is feasible as long as the composition of the brain (observation space, action space, neural net model) remains the same. The design of ml-agents might be a little tricky to set this up, because if a model is loaded for training, it will restore also some hyperparameters (e.g. current number of steps, which affects the learning rate). So it would be faster to start out with curriculum learning.

> What does the "IL" in the example scene names mean, does that stand for Imitiation Learning?

Yes, it is imitation learning.

> In what types of situations would you suggest to use_curiosity, and when not?...

Curiosity is meant to address the challenge of having delayed and spare rewards in environments.
https://arxiv.org/pdf/1705.05363.pdf

> Would it make sense to have a rayPerception type class but where it also collects the surface normals...

I usually write my own raycast logic instead of using the RayPerception class. My hint for you is to think of the observation space as the best detailed description of the environment. Maybe the environment is well described for the agent by using the position, normals... of the hit points

**Comment by JPhilipp** [comment]
> So that for example the cannon can see the bullet flying.

Makes sense, thank you!

> I usually write my own raycast logic instead of using the RayPerception
> class. My hint for you is to think of the observation space as the best
> detailed description of the environment.

Gotcha, thanks!
When no hit is found by the ray, do you just AddVectorObs(0f) to fill the required vector observation count?

**Comment by MarcoMeter** [comment]
> When no hit is found by the ray, do you just AddVectorObs(0f) to fill the required vector observation count?

If nothing is hit, I usually enter -1.


Here are some answers to other questions:

> Is anything on the roadmap for the future for training the setup values themselves via ML, e.g. an ML figuring...

Hyperparameter tuning (not meaning selecting properties for the observation space) is a challenge itself in machine learning. Usually, it is done based on experience and related problems. One could implement a script to test multiple setups automatically. However, having an AI optimizing hyperparameters to optimize an agent's behavior adds way more complexity.

> How do I pick the ideal rayPerception max ray distance?...

Debugging an agent is pretty tough. You always have to make sure that the information presented in the observation space is useful in helping the agent to achieve its goal. So it is really dependent on your environment.
